### PR TITLE
[SPARK-5307] SerializationDebugger to help debug NotSerializableException

### PR DESCRIPTION
**Pull request by rxin**
This patch adds a SerializationDebugger that is used to add more information to a NotSerializableException. When a NotSerializableException is encountered, the debugger tries to serialize the object one more time through a DebugStream that hooks into the internals of ObjectOutputStream to get the serialization stack. This ensures that there is no performance loss to run with SerializationDebugger, unlike setting the sun.io.serialization.extendedDebugInfo flag.

An example output looks like this:

```
org.apache.spark.serializer.NotSerializableClass
    Serialization stack (3):
    - org.apache.spark.serializer.NotSerializableClass@5e20dc10 (class org.apache.spark.serializer.NotSerializableClass)
    - org.apache.spark.serializer.SerializableClass2@521fb14e (class org.apache.spark.serializer.SerializableClass2)
    - org.apache.spark.serializer.SerializableClass1@5f54e92c (class org.apache.spark.serializer.SerializableClass1)
    Run the JVM with sun.io.serialization.extendedDebugInfo for more information.
```

When sun.io.serialization.extendedDebugInfo is turned on, this debugger no longer adds more information. Note that sun.io.serialization.extendedDebugInfo can show also the field name information, which is harder to get by the SerializationDebugger (technically possible with reflection but fairly convoluted).

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+      try {
+        out.writeObject(obj)
+        e
+      } catch {
+        case nse: NotSerializableException =>
+          new NotSerializableException(
+            nse.getMessage + "\n" +
+            s"\tSerialization stack (${out.stack.size}):\n" +
+            out.stack.map(o => s"\t- $o (class ${o.getClass.getName})").mkString("\n") + "\n" +
+            "\tRun the JVM with sun.io.serialization.extendedDebugInfo for more information.")
```
Comment (liancheng) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: Maybe `-Dsun.io.serialization.extendedDebugInfo`?

Comment (rxin) [inline core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala]: It is actually "-Dsun.io.serialization.extendedDebugInfo=true". Kinda long ...

**Comment by liancheng** [comment]
This LGTM.

**Comment by rxin** [comment]
Closing this one in favor of https://github.com/apache/spark/pull/4098

**Comment by geofflangenderfer** [comment]
could someone give a simple example of how to read the graph? I'm not sure where to start

**Pull request closed without merging**